table/tables/tables.go

```diff
@@ -454,8 +454,23 @@ func (t *TableCommon) rebuildIndices(ctx sessionctx.Context, txn kv.Transaction,
 		if t.meta.IsCommonHandle && idx.Meta().Primary {
 			continue
 		}
+		// when drop column with a index with it, the column and index will perform a combined schema change.
+		// column:   write-only  ->  delete-only  -> delete-reorg
+		// index:    write-only  ->  delete-only  -> delete-reorg
+		//
+		// under the latter two states, the update statement will not the fetch old column data up and write it
+		// back because drop column is not allowed to be cancelled in these states. But the state-combined index
+		// require the old column data to build the index key to conduct the delete operation. (update statement
+		// can only see the writable columns)
+		//
+		// Since drop column ddl can't be cancelled after stepping in these stage, previously we won't fulfil
+		// corresponding state column datum when we insert a record, hence, we shouldn't have to  clean the index
+		// record for it as well, which will be clean after the ddl is finished. Otherwise, we need to fetch
+		// delete-only/delete-reorg column's data up in update statement and remove the old index record here,
+		// which is not that elegant.
 		for _, ic := range idx.Meta().Columns {
-			if !touched[ic.Offset] {
+			// delete-only index column doesn't exist in the touched array.
+			if mysql.HasDropColumnWithIndexFlag(t.Columns[ic.Offset].Flag) || !touched[ic.Offset] {
```

Comment:
Can you just check its visibility? Like public, or delete org.

Reply:
We can't skip removing the index record because it may cause the data shown to be inconsistent.

Imagine that there are 2 schema versions in the same cluster:
```sql
alter table t drop column b; -- (running DDL)

create table t (a int, b int, c int, index idx(b)); -- TiDB 1
                       ^ write-only         ^ write-only

create table t (a int, b int, c int, index idx(b)); -- TiDB 2
                       ^ public             ^ public
```

Now TiDB 1 updates the row record. The row record is removed but the index record is not. This results in data inconsistency from the perspective of TiDB 2.

Reply:
yes, since the issue only error in the `delete-only` and `delete-reorg`, we should skip only in this two state.

Schema Co-Existence:
public         <->      write only  ---- update in `write only` still fetch the column up, so just refill it as what it was
write only   <->      delete only  ---- update in `delete only` will skip refilling, and dml in `write only` can't use the index either
delete only <->      delete reorg  --- update in `delete reorg` will skip refilling, and dml in `delete only` can't use the index either